client/src/modules/lead/EditLead/EditLeadHeader/EditLeadPipelinePopover/EditLeadPipelinePopover.tsx

```diff
@@ -0,0 +1,165 @@
+        target={this.props.target}
+        toggle={this.togglePopover}
+      >
+        <Card>
+          <CardBody className={styles.pipelineContainer}>
+            <div className={styles.pipelineHeader}>Pipeline</div>
+            <div className={styles.inputContainer}>
+              <Dropdown
+                isOpen={this.state.isDropdownOpen}
+                toggle={this.toggleDropdown}
+              >
+                <DropdownToggle className={styles.select}>
+                  {this.state.selectedFunnel.name}
+                </DropdownToggle>
+                {this.renderDropdownOptions()}
+              </Dropdown>
+            </div>
+            <SelectStageOnCreation 
+              stages={this.props.stages} 
+              onStageChange={this.onStageSelect}
+              title="Pipeline's stages"
+              stageId={this.state.selectedStageId}
+            />
+          </CardBody>
+          <CardFooter className={styles.buttons}>
+            <button onClick={this.onSave} className={styles.buttonSave}>
+              Save
+            </button>
+            <button className={styles.button} onClick={this.togglePopover}>
+              Cancel
+            </button>
+          </CardFooter>
+        </Card>
+      </Popover>
+    );
+  }
+
+  public componentWillMount() {
+    this.props.loadPipelinePopoverStages(this.props.lead.stage.funnel._id);    
+    this.setState({
+      selectedFunnel: this.props.lead.stage.funnel,
+      selectedStageId: this.props.lead.stage._id,
+    });
+  }
+
+  public componentWillReceiveProps(nextProps: Props) {
+    if (this.isFunnelChangedAndStagesLoaded(nextProps)) {
+      this.setState({
+        selectedStageId: nextProps.stages[0]._id,
+      });
+    } 
+
+    if (this.props.stages === nextProps.stages) {
+      this.onFunnelSelect(nextProps.lead.stage.funnel);
+    }
+  }
+
+  private isFunnelChangedAndStagesLoaded (props: Props): boolean {
+    return this.props.stages !== props.stages && this.props.stages.length !== 0;
+  }
+
+  private renderDropdownOptions() {
+    return (
+      <DropdownMenu>
+        {this.props.funnels.map((funnel: Funnel) => {
+        return (
+          <DropdownItem className={styles.option} onClick={this.onFunnelSelect.bind(this, funnel)} key={funnel._id}>
+            {funnel.name}
+          </DropdownItem>
+        );
+        })}
+      </DropdownMenu>
+    );
+  }
+
+  private onFunnelSelect(funnel: Funnel) {
+    this.setState({
+      selectedFunnel: funnel,
+    }, () => {
+      this.props.loadPipelinePopoverStages(this.state.selectedFunnel._id);
+    });
+  }
+
+  private onStageSelect = (stageId: string) => {
+    this.setState({
+      selectedStageId: stageId,
+    });
+  }
+
+  private toggleDropdown = () => {
+    this.setState((prevState: State) => ({
+      isDropdownOpen: !prevState.isDropdownOpen,
+    }));
+  }
+
+  private togglePopover = () => {    
+    this.setState({  
+      isDropdownOpen: false,
```

Comment:
these values are same as the initial state values, you should make `getInitialState()` and use it here and on state initilaization